test/mlir/onnx/onnx_canonicalization.mlir

```diff
@@ -1403,6 +1403,100 @@ func.func @expand_pow_into_constant(%arg0: tensor<3x4x5xf32>) -> tensor<3x4x5xf3
+// CHECK:           [[VAR_0_:%.+]] = "onnx.Pow"([[PARAM_0_]], [[PARAM_1_]]) : (tensor<3x4x5xf32>, tensor<1xi32>) -> tensor<3x4x5xf32>
+// CHECK:           onnx.Return [[VAR_0_]] : tensor<3x4x5xf32>
+// CHECK:         }
+}
+
+// -----
+
+// COM: conv introduced due to different size.
+func.func @pow_convert_size1(%arg0: tensor<3x4x5xf32>, %arg1: tensor<1xf64>) -> tensor<3x4x5xf32> {
+    %0 = "onnx.Pow"(%arg0, %arg1) : (tensor<3x4x5xf32>, tensor<1xf64>) -> tensor<3x4x5xf32>
+    onnx.Return %0 : tensor<3x4x5xf32>
+
+// mlir2FileCheck.py
+// CHECK-LABEL:  func.func @pow_convert_size1
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<3x4x5xf32>, [[PARAM_1_:%.+]]: tensor<1xf64>) -> tensor<3x4x5xf32> {
+// CHECK:           [[VAR_0_:%.+]] = "onnx.Cast"([[PARAM_1_]]) {saturate = 1 : si64, to = f32} : (tensor<1xf64>) -> tensor<1xf32>
+// CHECK:           [[VAR_1_:%.+]] = "onnx.Pow"([[PARAM_0_]], [[VAR_0_]]) : (tensor<3x4x5xf32>, tensor<1xf32>) -> tensor<3x4x5xf32>
+// CHECK:           onnx.Return [[VAR_1_]] : tensor<3x4x5xf32>
+// CHECK:         }
+}
+
+// -----
+
+// COM: conv introduced due to different size.
+func.func @pow_convert_size2(%arg0: tensor<3x4x5xi32>, %arg1: tensor<1xi64>) -> tensor<3x4x5xi32> {
+    %0 = "onnx.Pow"(%arg0, %arg1) : (tensor<3x4x5xi32>, tensor<1xi64>) -> tensor<3x4x5xi32>
+    onnx.Return %0 : tensor<3x4x5xi32>
+
+// mlir2FileCheck.py
+// CHECK-LABEL:  func.func @pow_convert_size2
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<3x4x5xi32>, [[PARAM_1_:%.+]]: tensor<1xi64>) -> tensor<3x4x5xi32> {
+// CHECK:           [[VAR_0_:%.+]] = "onnx.Cast"([[PARAM_1_]]) {saturate = 1 : si64, to = i32} : (tensor<1xi64>) -> tensor<1xi32>
+// CHECK:           [[VAR_1_:%.+]] = "onnx.Pow"([[PARAM_0_]], [[VAR_0_]]) : (tensor<3x4x5xi32>, tensor<1xi32>) -> tensor<3x4x5xi32>
+// CHECK:           onnx.Return [[VAR_1_]] : tensor<3x4x5xi32>
+// CHECK:         }
+}
+
+// -----
+
+// COM: conv introduced due to different size. Note it converts the i64 to f32... could have done i32 too.
+func.func @pow_convert_size3(%arg0: tensor<3x4x5xf32>, %arg1: tensor<1xi64>) -> tensor<3x4x5xf32> {
+    %0 = "onnx.Pow"(%arg0, %arg1) : (tensor<3x4x5xf32>, tensor<1xi64>) -> tensor<3x4x5xf32>
+    onnx.Return %0 : tensor<3x4x5xf32>
+
+// mlir2FileCheck.py
+// CHECK-LABEL:  func.func @pow_convert_size3
+// CHECK-SAME:   ([[PARAM_0_:%.+]]: tensor<3x4x5xf32>, [[PARAM_1_:%.+]]: tensor<1xi64>) -> tensor<3x4x5xf32> {
+// CHECK:           [[VAR_0_:%.+]] = "onnx.Cast"([[PARAM_1_]]) {saturate = 1 : si64, to = f32} : (tensor<1xi64>) -> tensor<1xf32>
+// CHECK:           [[VAR_1_:%.+]] = "onnx.Pow"([[PARAM_0_]], [[VAR_0_]]) : (tensor<3x4x5xf32>, tensor<1xf32>) -> tensor<3x4x5xf32>
+// CHECK:           onnx.Return [[VAR_1_]] : tensor<3x4x5xf32>
+// CHECK:         }
+}
+
+// -----
+
+// COM: conv introduced due to unsupported mlir combination (int ^ float)
+func.func @pow_convert_size4(%arg0: tensor<3x4x5xi32>, %arg1: tensor<1xf32>) -> tensor<3x4x5xi32> {
```

Comment:
This case `base i32, power f32 `is interesting. It seems that converting the base to `f32` would give a correct result. For example `4^(0.5)`, the answer is `2`, but if we cast `0.5` to  `0` or `1`, the answer is `1` or `4`.

Reply:
Interesting observation: it's kind of a strange case.... I asked the question on the slack onyx-operator channel.

Reply:
i wonder if the `int ^ float` case should be legal in the first place. seems like it's just not a well defined operation. after all there is no builtin MLIR op for this case anyway, so how would you lower it to krnl?

Reply:
> i wonder if the int ^ float case should be legal in the first place.

I tend to agree with that statement, as `sqrt `(aka `x ^ 0.5`) tends to give floating point numbers, that would then be rounded to an int? That sounds not that plausible. There is a discussion in the onnx/operator slack channel, feel free to provide your opinion there.

Maybe we could keep the current implementation as is at this time, @gramalingam seems to indicate that this policy is also the one of ORT. If/when the specs are clarified, we can then change it to the better defined standard.

What do you think @srcarroll @tungld ?

Reply:
If it is consistent with the one in ORT, I think it's fine to keep it as is. If we can put a memo on this case somewhere, it would be helpful in the future.

Reply:
Did not add an explicit "TODO", can add one in the next (unrelated) PR if you feel it would be a good thing.